### PR TITLE
fix: specify roottype in substrait fieldreference

### LINKED ISSUE #13645: Substrait: FieldReference not created correctly; missing a RootType
### Describe the bug

The FieldReference in Substrait is defined with `root_type` being `oneof`
```
  message FieldReference {

    // Whether this reference has an origin of a root struct or is based on the
    // ouput of an expression. When this is a RootReference and direct_reference
    // above is used, the direct_reference must be of a type StructField.
    oneof root_type {
      Expression expression = 3;
      RootReference root_reference = 4;
      OuterReference outer_reference = 5;
    }

```

A `RootReference` message is defined, that is empty but is used in most cases I've seen.

The code [here] (https://github.com/apache/datafusion/blob/fdb221fa645359b4bb65f600db1997ad6c865b65/datafusion/substrait/src/logical_plan/producer.rs#L2144) sets this to `None`.

When this is serialized to a protobuf and read by other engines (Substrait=Java in my case) this fails. with a 
` Unhandled type: ROOTTYPE_NOT_SET`

### To Reproduce

Create a DataFusion plan, and then try and consume it within say SubstraitJava

### Expected behavior

The protobuf should be valid and parse correctly 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by mbwhite**
## Which issue does this PR close?

Closes #13645 

## Rationale for this change

Should match the Substrait spec - None isn't valid for this protobuf

-->

## Are these changes tested?

Using the existing Substrait Tests; though I do note that some aren't passing any way

### CODE REVIEW

**Review by alamb — COMMENTED** [review]
Thank you @mbwhite  -- Would it be possible to write a test for this feature (so we don't break it in the future by accident)?

**Comment by mbwhite** [comment]
> Thank you @mbwhite -- Would it be possible to write a test for this feature (so we don't break it in the future by accident)?

Yes of course, if the basic function is good happy to proceed.

**Comment by alamb** [comment]
@vbarua or @Blizzara  as you are more experts in substrait, do you have a moment to review if this change is correct or not?

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 1]
```diff
@@ -2192,7 +2193,7 @@ fn substrait_field_ref(index: usize) -> Result<Expression> {
                     }),
                 )),
             })),
-            root_type: None,
+            root_type: Some(RootType::RootReference(RootReference{}) ),
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:2196]: Not necessary, but if you feel like it, could remove the duplication here: split this function into `substrait_field_ref` that returns the FieldReference and `substrait_field_ref_expr` that wraps it into Expression and then reusing `substrait_field_ref` in the `try_to_substrait_field_reference`.

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 2]
```diff
@@ -18,6 +18,7 @@
 use datafusion::config::ConfigOptions;
 use datafusion::optimizer::analyzer::expand_wildcard_rule::ExpandWildcardRule;
 use datafusion::optimizer::AnalyzerRule;
+use substrait::proto::expression::field_reference::{RootReference, RootType};
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:21]: nit: looks like the imports are kinda messy anyways but mind putting this somewhere next to other substrait things still?

Comment (mbwhite) [reply datafusion/substrait/src/logical_plan/producer.rs:21]: Sure;  I'd just let vscode sort out the imports; but yes can move it down.

File: datafusion/substrait/src/logical_plan/producer.rs (hunk 1 above)
Comment (mbwhite) [reply datafusion/substrait/src/logical_plan/producer.rs:2196]: Yes - I agree; though my rust is a few years out of date.. so getting up to speed again. So trying to make as few changes as needed...

File: datafusion/substrait/src/logical_plan/producer.rs [hunk 3]
```diff
@@ -2422,6 +2424,20 @@ mod test {
         Ok(())
     }
 
+    #[test]
+    fn to_field_reference() -> Result<()> {
+        let expression = substrait_field_ref(2)?;
+
+        match &expression.rex_type {
+            Some(RexType::Selection(field_ref)) => {
+                assert_ne!(field_ref.root_type, None);
```
Comment (Blizzara) [inline datafusion/substrait/src/logical_plan/producer.rs:2433]: would this work?
```suggestion
                assert_eq!(field_ref.root_type.expect("root type should be set"), RootType::RootReference(RootReference {}));
```

**Review by alamb — APPROVED** [review]
Thank you

File: datafusion/substrait/src/logical_plan/producer.rs (hunk 3 above)
Comment (mbwhite) [reply datafusion/substrait/src/logical_plan/producer.rs]: needs a slight change to pass the build

**Comment by alamb** [comment]
Sorry for the delay in merging this 🤦  Thanks again@

**Comment by mbwhite** [comment]
@alamb no problem;  any idea when the next release containing this might be?

**Pull request merged**